### ISSUE TITLE
The problem of recurrent small Re-ID datasets: VIPeR and GRID

### ISSUE THREAD

**Issue by wangli-upc** [KaiyangZhou/deep-person-reid#329]
Dear Zhou:  
        Your paper and framework are very graceful, thank you.  However I still get a problem:
        I read the paper “Learning Generalisable Omni-Scale Representations for Person Re-Identification” and used OSNet to recurrent the results of  VIPeR / GRID in this paper.  But there is a big gap between the results and this paper(more than 10% mAP/R1).
      What  parameter configuration is needed to use OSNet to recurrent the results of the paper？
      Could you share your configuration file to me?
      We appreciate very much for your time in answering my questions.

**Comment by KaiyangZhou** [comment]
to reproduce the results, you can follow the "implementation details" given in the paper

for viper and grid, different splits could lead to different performance so you need to follow the standard protocol to compute the average result over 10 splits, check this out https://kaiyangzhou.github.io/deep-person-reid/evaluation.html

**Comment by wangli-upc** [comment]
Dear Zhou:
        Firstly, we appreciate very much for your time to answer our question. you are so kind.
        I followed your advice and carefully checked the osnet parameter configuration and data splits method in the code.  Discover that the configuration in the code is the same as the parameter configuration in "implementation details" and the paper. See figure for details. But the result is not good. 
        Where are our parameters configured wrong? Could you share your configuration file to me? 
        I really appreciate your help.
        
![image](https://user-images.githubusercontent.com/34803397/78619086-dddb4980-78ae-11ea-9fb1-56accd2f0fd2.png)
![image](https://user-images.githubusercontent.com/34803397/78619097-e764b180-78ae-11ea-92e7-11880c7b308b.png)
![image](https://user-images.githubusercontent.com/34803397/78624332-f4d46880-78bb-11ea-8f86-ad45574e9c64.png)
![image](https://user-images.githubusercontent.com/34803397/78624356-fc940d00-78bb-11ea-87cf-48ccb2609c58.png)


**Parameter configuration:**  
model:
  name: 'osnet_x1_0'
  pretrained: True

data:
  type: 'image'
  sources: ['viper']
  targets: ['viper']
  height: 256
  width: 128
  combineall: False
  transforms: ['random_flip', 'random_erase']
  save_dir: 'log/osnet/osnet_pretrain_viper'

loss:
  name: 'softmax'
  softmax:
    label_smooth: True

train:
  optim: 'amsgrad'
  lr: 0.0015
  max_epoch: 250
  batch_size: 64
  fixbase_epoch: 10
  open_layers: ['classifier']
  lr_scheduler: 'cosine'

test:
  batch_size: 300
  dist_metric: 'euclidean'
  normalize_feature: False
  evaluate: False
  start_eval: 200
  eval_freq: 1
  rerank: False
  visactmap: False

**Comment by KaiyangZhou** [comment]
oops, my bad, the training configs on small datasets, i.e. viper and grid, were different, I forgot mentioning this

try this `max_epoch =30, fixbase_epoch=5, stepsize=20` (not sure)

**Comment by wangli-upc** [comment]
Dear Zhou:
      Thanks very very much for your answer. 
      We modified the configuration file according to your suggestion and did an experiment, but the result is still not good (osnet  R1/map = 22.5%/34.1%).
      Where are our parameters configured wrong? Could you give us a more guidance on our configuration?
      I really appreciate your help.
![image](https://user-images.githubusercontent.com/34803397/78794941-f484bc00-79e6-11ea-9d1b-3746d4ab8ceb.png)

my configuration parameters are as follows: 

model:
  name: 'osnet_x1_0'
  pretrained: True

data:
  type: 'image'
  sources: ['grid']
  targets: ['grid']
  height: 256
  width: 128
  combineall: False
  transforms: ['random_flip', 'random_erase']
  save_dir: 'log/osnet/osnet_pretrain_grid'

loss:
  name: 'softmax'
  softmax:
    label_smooth: True

train:
  optim: 'amsgrad'
  lr: 0.0015
  max_epoch: 30
  batch_size: 64
  fixbase_epoch: 5
  open_layers: ['classifier']
  lr_scheduler: 'multi_step'
  stepsize: [20]

test:
  batch_size: 300
  dist_metric: 'euclidean'
  normalize_feature: False
  evaluate: False
  start_eval: 0
  eval_freq: 1
  rerank: False
  visactmap: False

**Comment by KaiyangZhou** [comment]
that's very weird to me, can you provide the full log?

can you check if you 100% follow the steps shown in the paper, like pretraining on big datasets, using proper fine-tuning techniques, label smoothing, blah blah

**Comment by wangli-upc** [comment]
Dear Zhou:
        Thanks very much, Your suggestion helped us a lot, and the accuracy has been improved to a certain extent.  
        grid dataset -> R1= 24.8%.
        However, results in the paper  “Learning Generalisable Omni-Scale Representations for Person Re-Identification”  is  ->  R1= 38.2%
        There is also a gap in R1 accuracy in viper.   R1=56.6%(our experiment) --> R1=68%(in paper).
         We use the MSMT database for pre-training，and fine-tuning on the grid/viper.
        Where are our parameters configured wrong? Could you give us a more guidance on our configuration?
         I really really appreciate your help.  my email: li.wang.upc@foxmail.com 
         Our configuration file and results are as follows:

**configuration**
adam:
  beta1: 0.9
  beta2: 0.999
cuhk03:
  classic_split: False
  labeled_images: False
  use_metric_cuhk03: False
data:
  combineall: False
  height: 256
  load_train_targets: False
  norm_mean: [0.485, 0.456, 0.406]
  norm_std: [0.229, 0.224, 0.225]
  root: /home/datasets
  save_dir: log/osnet/osnet_pretrain_grid
  sources: ['grid']
  split_id: 0
  targets: ['grid']
  transforms: ['random_flip', 'random_erase']
  type: image
  width: 128
  workers: 4
engine:
  name: default
loss:
  members: ['out_x']
  name: softmax
  softmax:
    label_smooth: True
  triplet:
    margin: 0.3
    weight_t: 0.5
    weight_x: 1.0
market1501:
  use_500k_distractors: False
model:
  load_weights: 
  name: osnet_x1_0
  pretrained: True
  resume: 
rmsprop:
  alpha: 0.99
sampler:
  mode: single
  multi_sampler: ['RandomSampler', 'RandomIdentitySampler']
  num_instances: 4
  switch_epoch: 100
  train_sampler: RandomSampler
sgd:
  dampening: 0.0
  momentum: 0.9
  nesterov: False
test:
  batch_size: 300
  dist_metric: euclidean
  eval_freq: -1
  evaluate: False
  normalize_feature: False
  ranks: [1, 5, 10, 20]
  rerank: False
  start_eval: 0
  visactmap: False
  visrank: False
  visrank_topk: 10
train:
  base_lr_mult: 0.1
  batch_size: 64
  fixbase_epoch: 5
  gamma: 0.1
  lr: 0.0015
  lr_scheduler: cosine
  max_epoch: 30
  new_layers: ['classifier']
  open_layers: ['classifier']
  optim: amsgrad
  print_freq: 20
  seed: 1
  staged_lr: False
  start_epoch: 0
  stepsize: [20]
  weight_decay: 0.0005
use_gpu: True
video:
  pooling_method: avg
  sample_method: evenly
  seq_len: 15

Collecting env info ...
** System info **
PyTorch version: 1.4.0+cu100
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: CentOS Linux release 7.6.1810 (Core) 
GCC version: (GCC) 5.3.1 20160406 (Red Hat 5.3.1-6)
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB

Nvidia driver version: 410.104
cuDNN version: /usr/lib64/libcudnn.so.7.6.3

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.4.0+cu100
[pip3] torchfile==0.1.0
[pip3] torchnet==0.0.4
[pip3] torchreid==1.0.9
[pip3] torchvision==0.5.0+cu100
[conda] torch                     1.4.0+cu100               <pip>
[conda] torchfile                 0.1.0                     <pip>
[conda] torchnet                  0.0.4                     <pip>
[conda] torchreid                 1.0.9                     <pip>
[conda] torchvision               0.5.0+cu100               <pip>
        Pillow (7.0.0)

Building train transforms ...
+ resize to 256x128
+ random flip
+ to torch tensor of range [0, 1]
+ normalization (mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
+ random erase
Building test transforms ...
+ resize to 256x128
+ to torch tensor of range [0, 1]
+ normalization (mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
=> Loading train (source) dataset
=> Loaded GRID
  ----------------------------------------
  subset   | # ids | # images | # cameras
  ----------------------------------------
  train    |   125 |      250 |         6
  query    |   125 |      125 |         5
  gallery  |   126 |      900 |         5
  ----------------------------------------
=> Loading test (target) dataset
=> Loaded GRID
  ----------------------------------------
  subset   | # ids | # images | # cameras
  ----------------------------------------
  train    |   125 |      250 |         6
  query    |   125 |      125 |         5
  gallery  |   126 |      900 |         5
  ----------------------------------------


  **************** Summary ****************
  source            : ['grid']
  # source datasets : 1
  # source ids      : 125
  # source images   : 250
  # source cameras  : 6
  target            : ['grid']
  *****************************************


Building model: osnet_x1_0
Successfully loaded imagenet pretrained weights from "/home/fanbaoyu/.cache/torch/checkpoints/osnet_x1_0_imagenet.pth"
** The following layers are discarded due to unmatched keys or layer size: ['classifier.weight', 'classifier.bias']
Model complexity: params=2,193,616 flops=978,878,352
Building softmax-engine for image-reid
=> Start training
* Only train ['classifier'] (epoch: 1/5)
* Only train ['classifier'] (epoch: 2/5)
* Only train ['classifier'] (epoch: 3/5)
* Only train ['classifier'] (epoch: 4/5)
* Only train ['classifier'] (epoch: 5/5)
=> Final test
##### Evaluating grid (source) #####
Extracting features from query set ...
Done, obtained 125-by-512 matrix
Extracting features from gallery set ...
Done, obtained 900-by-512 matrix
Speed: 0.3986 sec/batch
Computing distance matrix with metric=euclidean ...
Computing CMC and mAP ...
** Results **
mAP: 34.7%
CMC curve
Rank-1  : 24.8%
Rank-5  : 43.2%
Rank-10 : 51.2%
Rank-20 : 57.6%
##### Evaluating grid (source) #####
Extracting features from query set ...
Done, obtained 125-by-512 matrix
Extracting features from gallery set ...
Done, obtained 900-by-512 matrix
Speed: 0.0306 sec/batch
Computing distance matrix with metric=cosine ...
Computing CMC and mAP ...
** Results **
**mAP: 37.6%**
CMC curve
**Rank-1  : 28.8%**
Rank-5  : 45.6%
Rank-10 : 53.6%
Rank-20 : 61.6%
Checkpoint saved to "log/osnet/osnet_pretrain_grid/model.pth.tar-30"
Elapsed 0:00:42
